R/data-ddl-utils.R

```diff
@@ -0,0 +1,99 @@
+#' Function runs the `code`, masks the `code` and creates `teal_data` object.
+#' @param data (`teal_data`) object
+#' @param code (`language`) code to evaluate
+#' @param input (`list`) containing inputs to be used in the `code`
+#' @param input_mask (`list`) containing inputs to be masked in the `code`
+#'
+#' @return `teal_data` object
+#'
+#' @export
+eval_and_mask <- function(data,
+                          code,
+                          input = list(),
+                          input_mask = list()) {
+  # todo: do we need also within_and_mask?
+  checkmate::assert_list(input)
+  if (inherits(input, "reactivevalues")) {
+    input <- shiny::reactiveValuesToList(input)
+  }
+  # evaluate code and substitute input
+  data <- teal.code::eval_code(data, .substitute_code(code, args = input))
+  if (inherits(data, "qenv.error")) {
+    return(data)
+  }
+
+  if (identical(ls(data@env), character(0))) {
+    warning(
+      "Evaluation of `ddl` code haven't created any objects.\n",
+      "Please make sure that the code is syntactically correct and creates necessary data."
+    )
+  }
+
+  if (!missing(input_mask)) {
+    # mask dynamic inputs with mask
+    input <- utils::modifyList(input, input_mask)
+
+    # replace last code entry with masked code
+    # format_expression needed to convert expression into character(1)
+    #  question: warnings and errors are not masked, is it ok?
+    data@code[length(data@code)] <- format_expression(.substitute_code(code, args = input))
+  }
+
+  # todo: should it be here or in datanames(data)?
+  if (length(datanames(data)) == 0) {
+    datanames(data) <- ls(data@env)
+  }
+
+  data
+}
+
+#' Substitute symbols in the code
+#'
+#' Function replaces symbols in the provided code by values of the `args` argument.
+#'
+#' @param code (`language`) code to substitute
+#' @param args (`list`) named list or arguments
+#' @keywords internal
+.substitute_code <- function(code, args) {
+  do.call(
+    substitute,
+    list(
+      expr = do.call(
+        substitute,
+        list(expr = code)
+      ),
+      env = args
+    )
+  )
+}
+
+#' Convenience wrapper for ddl
+#' @export # todo: do we want to export this?
+ddl <- function(code, input_mask, ui, server) {
```

Comment:
Could be used in this way 

<details>
<summary> app code </summary>
 
```r
options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
library(scda)
pkgload::load_all("teal")

ui <- function(id) {
  ns <- NS(id)
  tagList(
    textInput(ns("username"), label = "Username"),
    passwordInput(ns("password"), label = "Password"),
    actionButton(ns("submit"), label = "Submit")
  )
}

server <- function(id, code, input_mask) {
  moduleServer(id, function(input, output, session) {
    eventReactive(input$submit, {
      masked_data <- teal_data() |> eval_and_mask(code = code, input = input, input_mask = input_mask)
    })
  })
}

data <- ddl(
  code = quote({
    open_conn(username = username, password = password)
    ADSL <- scda::synthetic_cdisc_data("latest")$adsl
    ADTTE <- scda::synthetic_cdisc_data("latest")$adtte
  }),
  input_mask = list(
    username = quote(askpass::askpass()),
    password = quote(askpass::askpass())
  ),
  ui = ui,
  server = server
)

app <- init(
  data = data,
  modules = modules(
    teal.modules.general::tm_data_table(label = "yolo")
  ),
  filter = teal_slices(
    teal_slice(dataname = "ADSL", varname = "AGE", selected = c(18, 60)),
    teal_slice(dataname = "inexisting", varname = "AGE")
  )
)

runApp(app)
```

</details>

Reply:
Question: Do we actually require the function ddl()? It only utilize `delayed_data` with an additional argument . Perhaps it would be more suitable as a additional example in documentation or  in the vignette?

Reply:
Exactly, I think the same

Reply:
> We can export delayed_data (name to decide) and expose "delayed data loading" in the documentation as an extension of delayed_data

my preference would be exporting `delayed_data`  as it might be more helpful to understand the class and concept itself to user.

Reply:
> Not sure why I can not see my ques above so pasting  here again,

> Question: Do we require the function ddl()? It only utilize `delayed_data` with an additional argument . Perhaps it would be more suitable as additonal example in documentation or  in the vignette?

Reply:
> my preference would be exporting delayed_data as it might be more helpful to understand the class and concept itself to user.

I have the same opinion. Besides, `delayed_data` might have much wider application

Reply:
I don't think exposing makes it easier to understand as there is `server_formals`, `extra_args`, `server_args`, `extra_formals`. If app developer understand that much of R and Shiny, would he be able to craft his own tool for his needs, instead of using teal? Even though I vote to expose

Reply:
`extra_args` and `extra_formals` are just helpful to print relevant error message. There are `server` formals and `...` which need to be matched. We need to throw meaningful message to alert that they don't match ;]

Reply:
I think dark smog clouds has overcome my brain today :)